File: 백준/Gold/16563. 어려운 소인수분해/어려운 소인수분해.py
def sieve_of_eratosthenes(n):
    prime = [True for _ in range(n+1)]
    p = 2
    while (p**2 <= n):
        if (prime[p] == True):
            for i in range(p**2, n+1, p):
                prime[i] = False
        p += 1
    return [p for p in range(2, n+1) if prime[p]]

primes = sieve_of_eratosthenes(int(5000000**0.5))
for num in map(int, open(0).read().split()[1:]):
    factors = []
    for prime in primes:
        if prime * prime > num:
            break
        while (num % prime == 0):
            factors.append(prime)
            num //= prime
    if num > 1:
        factors.append(num)
    print(*factors)

File: 백준/Gold/16563. 어려운 소인수분해/어려운 소인수분해.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> sieve_of_eratosthenes(int n) {
    vector<bool> prime(n + 1, true);
    vector<int> res;
    for (int p = 2; p * p <= n; p++) {
        if (prime[p]) {
            for (int i = p * p; i <= n; i += p)
                prime[i] = false;
        }
    }
    for (int p = 2; p <= n; p++) {
        if (prime[p]) res.push_back(p);
    }
    return res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    vector<int> primes = sieve_of_eratosthenes((int)sqrt(5000000));
    for (int num : nums) {
        vector<int> factors;
        for (int prime : primes) {
            if ((long long)prime * prime > num) break;
            while (num % prime == 0) {
                factors.push_back(prime);
                num /= prime;
            }
        }
        if (num > 1) factors.push_back(num);
        for (int i = 0; i < (int)factors.size(); i++) {
            if (i) cout << " ";
            cout << factors[i];
        }
        cout << "\n";
    }
}